Ensure that `pointerUp` Action is Being Issued on Element Being Tested

dshin-moz opened the pull request:
There does not seem to be any guarantee that the pointer will move between actions.

Fixes web-platform-tests/interop#606 - verified on local Chromium build.

dshin-moz committed:
Ensure that `pointerUp` Action is Being Issued on Element Being Tested

There does not seem to be any guarantee that the pointer will move between actions.

foolip commented:
@dshin-moz it looks like this breaks the test in Firefox instead:
https://wpt.fyi/results/css/selectors/invalidation/user-action-pseudo-classes-in-has.html?label=pr_head&max-count=1&pr=43214

Did you intend for that to happen?

dshin-moz commented:
Hm.. That's unfortunate - Looking at the logs, the failures happen before the changed line. Checking locally  with `--verify-repeat-restart=500` and it happens before this change as well, it's just that the failure rate is 3/500 :/

dshin-moz committed:
Update user-action-pseudo-classes-in-has.html

* `pointerMove` origins set on elements instead of viewport + clientRect location
* Specifically wait on promise fulfilled microtask being queued after relevant events are emitted 
* Break down the tests so that one test don't rely on the pointer hovering over where the previous test put it

dshin-moz committed:
Update user-action-pseudo-classes-in-has.html

No need to depend on event - Dispatched actions are [guaranteed](https://www.w3.org/TR/webdriver2/#dfn-dispatch-actions-inner) to resolve after DOM events generated by actions are processed.

foolip commented on css/selectors/invalidation/user-action-pseudo-classes-in-has.html:
```diff
@@ -37,47 +38,56 @@
-                  "subject3 should be navy again");
+        .Actions()
+        .pointerUp()
+        .pointerMove(0, 0, {origin: unhoverme})
```
This removes coverage for what was previously tested, which was to just release the mouse button without moving it. Can you keep this part of the test?

```suggestion
        .pointerMove(0, 0, {origin: hoverme})
```

And keep the assertions as they were. If you think it's also useful to test moving the mouse right after mouseup, additional steps for that would be OK, but I'm not sure I see the difference to the first unhoverme bits.

dshin-moz replied:
It does remove that coverage - but does the suggested change really test "going from `:hover:active` to `:hover`?"
I could be misreading this, but the current & suggested action sequence decomposes into two ticks, right? So it'd be a succession of "`:has(:hover:active)` invalidates" (Which is similar to the current change), then "`:has(:hover)` invalidates" (Which is already tested)?

dshin-moz commented on css/selectors/invalidation/user-action-pseudo-classes-in-has.html:
```diff
@@ -37,47 +38,56 @@
-                  "subject1 should be blue again");
-    assert_equals(getComputedStyle(subject3).color, "rgb(0, 0, 128)",
-                  "subject3 should be navy again");
+        .Actions()
+        .pointerUp()
+        .pointerMove(0, 0, {origin: unhoverme})
```
Could do:
```suggestion
        .Actions()
        .pointerUp()
        .pointerMove(0, 0, {origin: unhoverme})

    // Independent action sequence that should have the last invalidation to be `:has(:hover:active)` -> `:has(:hover)`
    await new test_driver
        .Actions()
        .pointerMove(0, 0, {origin: hoverme})
        .pointerDown()
        .pointerUp()
        .send();
    // ...
```

foolip replied:
Now that I understand the motivation for the change, I don't think it's important to keep testing the old thing. I think we should treat `:active` and `:hover` as separate, and both are now testing in both directions (matching to not matching and vice versa)

foolip replied:
I think testing `:has(:hover:active)` &rarr; `:has(:hover)` is perhaps not needed, but you already wrote the test, so let's have it! (I deleted a previous comment saying we don't need this.)

dshin-moz committed:
Update user-action-pseudo-classes-in-has.html

Keep the test that invalidates from `:has(:hover:active)` -> `:has(:hover)`.
